src/main/java/spireCafe/util/CafeUtil.java

```diff
@@ -0,0 +1,143 @@
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import spireCafe.Anniv7Mod;
+import spireCafe.abstracts.AbstractAttraction;
+import spireCafe.abstracts.AbstractBartender;
+import spireCafe.abstracts.AbstractCafeInteractable;
+import spireCafe.abstracts.AbstractMerchant;
+import spireCafe.abstracts.AbstractPatron;
+import spireCafe.abstracts.AbstractCafeInteractable.FacingDirection;
+
+public class CafeUtil {
+
+    public static HashMap<Class<? extends AbstractCafeInteractable>, Method> canSpawnMethods = new HashMap<>();
+
+    public static HashMap<String, Class<? extends AbstractCafeInteractable>> getAllInteractableByType(Class<? extends AbstractCafeInteractable> clz) {
+        return Anniv7Mod.interactableClasses.entrySet().stream()
+        .filter(e -> clz.isAssignableFrom(e.getValue()))
+        .collect(Collectors.toMap(
+            e -> e.getKey(),
+            e -> e.getValue(),
+            (oldValue, newValue) -> oldValue,
+            HashMap::new));
+    }
+    
+    public static HashMap<String, Class<? extends AbstractCafeInteractable>> getAllAttractions() {
+         return getAllInteractableByType(AbstractAttraction.class);
+    }
+    public static HashMap<String, Class<? extends AbstractCafeInteractable>> getAllBartenders() {
+         return getAllInteractableByType(AbstractBartender.class);
+    }
+    public static HashMap<String, Class<? extends AbstractCafeInteractable>> getAllMerchants() {
+         return getAllInteractableByType(AbstractMerchant.class);
+    }
+    public static HashMap<String, Class<? extends AbstractCafeInteractable>> getAllPatrons() {
+         return getAllInteractableByType(AbstractPatron.class);
+    }
+
+    private static boolean canSpawn(Class<? extends AbstractCafeInteractable> clz) {
+        Method canSpawnMethod = null;
+        if (canSpawnMethods.containsKey(clz)) {
+            canSpawnMethod = canSpawnMethods.get(clz);
+        }
+        else {
+            Method[] methods = clz.getDeclaredMethods();
+            for (Method m : methods) {
+                if (Modifier.isStatic(m.getModifiers()) && m.getName().equals("canSpawn") && m.getReturnType().equals(boolean.class) && m.getParameterCount() == 0) {
+                    m.setAccessible(true);
+                    canSpawnMethod = m;
+                    break;
+                }
+            }
+            canSpawnMethods.put(clz, canSpawnMethod);
+        }
+        try {
+            return canSpawnMethod == null || (boolean)canSpawnMethod.invoke(null);
+        } catch (IllegalAccessException | InvocationTargetException e) {
+            throw new RuntimeException(e);
+        }
+    }
+    
+    public static HashMap<String, Class<? extends AbstractCafeInteractable>> getAllValidInteractableByType(Class<? extends AbstractCafeInteractable> clz) {
```

Comment:
Similar to the comment above, there's enough logic here that it would be best if it wasn't duplicated.

It looks to me like this is based on `CafeRoom.getPossibilities`, with the difference being that it returns a HashMap instead of a List. But it's easy to resolve that: you can put this function into `CafeRoom`, and change `getPossibilities` to call this function and then take only the values and return a list. It adds an "unnecessary" HashMap creation to the process, but in practice that doesn't matter at all (nobody will ever notice the performance difference).

(The extra operation could also be avoided by making a method that just does the part up to the canSpawn step and returns a `Stream<...>`. Feel free to do that instead if you want.)